gcloud-java-dns/src/main/java/com/google/gcloud/dns/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1360 @@
+        }
+      };
+      thread.start();
+      return thread;
+    } else {
+      applyExistingChange(projectId, zoneName, changeId);
+      return null;
+    }
+  }
+
+  /**
+   * Applies changes to a zone. Repeatedly tries until succeeds. Thread safe and deadlock safe.
+   */
+  private void applyExistingChange(String projectId, String zoneName, String changeId) {
+    Change change = findChange(projectId, zoneName, changeId);
+    if (change == null) {
+      return; // no such change exists, nothing to do
+    }
+    ZoneContainer wrapper = findZone(projectId, zoneName);
+    if (wrapper == null) {
+      return; // no such zone exists; it might have been deleted by another thread
+    }
+    AtomicReference<ImmutableList<RrsetWrapper>> dnsRecords = wrapper.dnsRecords();
+    while (true) {
+      // managed zone must have a set of records which is not null
+      ImmutableList<RrsetWrapper> original = dnsRecords.get();
+      List<RrsetWrapper> copy = Lists.newLinkedList(original);
+      // apply deletions first
+      List<ResourceRecordSet> deletions = change.getDeletions();
+      if (deletions != null) {
+        List<RrsetWrapper> transformedDeletions = Lists.transform(deletions,
+            RrsetWrapper.WRAP_FUNCTION);
+        copy.removeAll(transformedDeletions);
+      }
+      // apply additions
+      List<ResourceRecordSet> additions = change.getAdditions();
+      if (additions != null) {
+        for (ResourceRecordSet addition : additions) {
+          String id = getUniqueId(copy);
+          RrsetWrapper rrsetWrapper = new RrsetWrapper(addition);
+          rrsetWrapper.setId(id);
+          copy.add(rrsetWrapper);
+        }
+      }
+      // make it immutable and replace
+//      boolean success = dnsRecords.replace(zoneName, original, ImmutableList.copyOf(copy));
+      boolean success = dnsRecords.compareAndSet(original, ImmutableList.copyOf(copy));
+      if (success) {
+        break; // success if no other thread modified the value in the meantime
+      }
+    }
+    // set status to done
+    change.setStatus("done");
+  }
+
+  /**
+   * Lists zones. Next page token is the last listed zone name and is returned only of there is more
+   * to list.
+   */
+  Response listZones(String projectId, Map<String, Object> options) {
+    Response response = checkListOptions(options);
+    if (response != null) {
+      return response;
+    }
+    ConcurrentSkipListMap<String, ZoneContainer> containers = findProject(projectId).zones();
+    String[] fields = (String[]) options.get("fields");
+    String dnsName = (String) options.get("dnsName");
+    String pageToken = (String) options.get("pageToken");
+    Integer maxResults = options.get("maxResults") == null
+        ? null : Integer.valueOf((String) options.get("maxResults"));
+    // matches will be included in the result if true
+    boolean listing = (pageToken == null || !containers.containsKey(pageToken));
```

Comment:
Did you see my response in the listing vs tailMap thread?  It's the 3rd compressed comment, but I'll copy here for convenience:

In response to: "I tried and I prefer the variable. When using tailMap, you still need to check if you should list all values (because pageToken is not provided or it does not exist in the map) or only the one after the pageToken. Furthermore, if you are listing with page token, you have to skip the first value of the tailMap unlike when listing all the results. Thus, it just makes one flag variable replaced by another and one boolean update replaced by condition test, and does not reduce the complexity of the code."

I wrote:
"You can choose to make the tailMap non-inclusive. I agree that you need to check whether to iterate over all values or just the tail map once before the loop, but I think this is better than needlessly always iterating over the entire map and checking whether to include the result in each iteration."